### PR TITLE
Feature/20789 softdelete pupil reason check

### PR DESCRIPTION
**Pull request by danwilliams**
There was no `isDeleted` column in the database for soft-deleting, so I also created a migration id 81 that adds it.

### CODE REVIEW

File: admin/services/data-access/pupil.data.service.js [hunk 1]
```diff
@@ -299,7 +299,7 @@ pupilDataService.sqlFindSortedPupilsWithAttendanceReasons = async (dfeNumber, so
   SELECT p.*, pg.group_id, ac.reason
   FROM ${sqlService.adminSchema}.${table} p 
     INNER JOIN ${sqlService.adminSchema}.[school] s ON p.school_id = s.id
-    LEFT OUTER JOIN ${sqlService.adminSchema}.[pupilAttendance] pa ON p.id = pa.pupil_id 
+    LEFT OUTER JOIN ${sqlService.adminSchema}.[pupilAttendance] pa ON p.id = pa.pupil_id AND (pa.isDeleted IS NULL OR pa.isDeleted = 0)
```
Comment (GeorgeChatzigiannis) [inline admin/services/data-access/pupil.data.service.js]: If the column is not null shouldn’t it only include the ones equal to 0?

**Comment by GuyHarwood** [comment]
@GeorgeChatzigiannis does this soft delete flag affect any of the pupil status calculations?

**Comment by GeorgeChatzigiannis** [comment]
@GuyHarwood that's a very good point.  
@danwilliams The pupil status service will definitely require a change to accommodate the pupil attendance soft deletion since its looking at records only based on pupil id.

**Comment by GuyHarwood** [comment]
It failed CI, so i hope thats because the automated tests around the raised concern were the ones that failed!

**Comment by danwilliams** [comment]
@GeorgeChatzigiannis it doesn’t affect the pupil status because that service is using the methods in pupil-attendance.data.service to check for id, and I’ve changed them (`findByPupilIds` and `findOneByPupilId`) to check for isDeleted=0 too so that should be handled. Are you happy with this? Or have I missed anything? 😄

**Comment by GeorgeChatzigiannis** [comment]
@danwilliams you are correct, pupil status service should work as intended with these changes.

What about the change I am proposing for pupil data service?

Under which case should we expect the value to be NULL?

**Comment by danwilliams** [comment]
@GeorgeChatzigiannis I think it's supposed to accept nulls too, that’s why it’s a `LEFT OUTER JOIN`, because we need to get all users (that have no reasons, or that already had a reason assigned) in order to display them in the list where reasons can be assigned to users.

**Pull request merged**